compiler/fir/checkers/src/org/jetbrains/kotlin/fir/analysis/checkers/declaration/FirPropertyInitializationChecker.kt

```diff
@@ -7,76 +7,50 @@ package org.jetbrains.kotlin.fir.analysis.checkers.declaration
 
 object FirPropertyInitializationChecker : FirRegularClassChecker() {
     override fun check(declaration: FirRegularClass, context: CheckerContext, reporter: DiagnosticReporter) {
-        val properties = mutableSetOf<FirPropertySymbol>()
-        val toReport = mutableSetOf<FirVariableAssignment>()
-        for (decl in declaration.declarations.asReversed()) {
-            when (decl) {
-                is FirProperty -> properties.add(decl.symbol)
-                is FirAnonymousInitializer ->
-                    if (properties.isNotEmpty()) {
-                        collectInitAssignments(decl, properties, toReport)
-                    }
-                else -> {
-                }
-            }
-        }
-        toReport.forEach { exp ->
-            val propertySymbol = exp.lValue.toResolvedCallableSymbol() ?: return@forEach
-            reporter.reportOn(exp.lValue.source, FirErrors.INITIALIZATION_BEFORE_DECLARATION, propertySymbol, context)
-        }
-    }
+        val interestingProperties = mutableSetOf<FirPropertySymbol>()
+        val visitor = object : FirVisitorVoid() {
+            override fun visitElement(element: FirElement) = element.acceptChildren(this)
 
-    private fun collectInitAssignments(
-        init: FirAnonymousInitializer,
-        properties: Properties,
-        assignments: MutableSet<FirVariableAssignment>
-    ) {
-        val graph = init.controlFlowGraphReference?.controlFlowGraph ?: return
-        val visitor = AssignmentVisitor(properties) { assignments.add(it) }
-        graph.collectDataForNode(TraverseDirection.Forward, emptySet(), visitor, visitSubGraphs = false)
-    }
+            override fun visitConstructor(constructor: FirConstructor) {}
 
-    private class AssignmentVisitor(private val interestingProperties: Properties, private val acceptor: (FirVariableAssignment) -> Unit) :
-        ControlFlowGraphVisitor<Properties, Collection<Pair<EdgeLabel, Properties>>>() {
+            override fun visitSimpleFunction(simpleFunction: FirSimpleFunction) {}
 
-        override fun visitNode(node: CFGNode<*>, data: Collection<Pair<EdgeLabel, Properties>>): Properties {
-            if (data.isEmpty()) return emptySet()
-            return data.map { it.second }.reduce { l, r -> l.intersect(r) }
-        }
+            override fun visitPropertyAccessor(propertyAccessor: FirPropertyAccessor) {}
 
-        override fun <T> visitUnionNode(
-            node: T, data: Collection<Pair<EdgeLabel, Properties>>
-        ): Properties where T : CFGNode<*>, T : UnionNodeMarker {
-            if (data.isEmpty()) return emptySet()
-            return data.map { it.second }.reduce { l, r -> l + r }
+            override fun visitAnonymousFunction(anonymousFunction: FirAnonymousFunction) {
+                // TODO? if (anonymousFunction.invocationKind.isInPlace) visitElement(anonymousFunction)
+            }
+
+            override fun visitRegularClass(regularClass: FirRegularClass) {}
```

Comment:
What about anonymous objects here? Looks like we need to visit `init` blocks and property initializers of them too
```kotlin
class A {
    init {
        object {
            init {
                x = 10 // INITIALIZATION_BEFORE_DECLARATION
            }
            
            fun foo() {
                x = 10 // VAL_REASSIGNMENT
            }
        }
    }

    val x: Int
}
```

Reply:
Anonymous objects go through visitAnonymousObject -> visitElement -> acceptChildren, so the error in `init` should be reported IIRC

Reply:
Looks like that indeed